### PR TITLE
Stream API Refactoring

### PR DESCRIPTION
**Pull request by aljoscha**
This is a WIP of the refactoring. I still want to add Javadocs and a new join operator based on tagged union that uses the new windowing operators.

### CODE REVIEW

**Comment by aljoscha** [comment]
I updates this to also add Javadocs for the new windowing semantics/internals.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/datastream/IterativeDataStream.java [hunk 1]
```diff
@@ -81,56 +81,56 @@ protected IterativeDataStream(DataStream<T> dataStream, long maxWaitTime) {
...
 	 * For type safety the user needs to define the feedback type
 	 * 
 	 * @param feedbackTypeClass
 	 *            Class of the elements in the feedback stream.
-	 * @return A {@link ConnectedIterativeDataStream}.
+	 * @return A {@link ConnectedIterativeDataStreams}.
 	 */
-	public <F> ConnectedIterativeDataStream<T, F> withFeedbackType(Class<F> feedbackTypeClass) {
+	public <F> ConnectedIterativeDataStreams<T, F> withFeedbackType(Class<F> feedbackTypeClass) {
 		return withFeedbackType(TypeExtractor.getForClass(feedbackTypeClass));
```
Comment (ktzoumas) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/datastream/IterativeDataStream.java]: Why ConnectedIterativeDataStreams and not ConnectedIterativeStreams (following the naming of the other classes)?

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java [hunk 2]
```diff
@@ -24,49 +24,169 @@
...
+	 *
 	 * @param reducer
 	 *            The {@link ReduceFunction} that will be called for every
 	 *            element of the input values with the same key.
 	 * @return The transformed DataStream.
 	 */
 	public SingleOutputStreamOperator<T, ?> reduce(ReduceFunction<T> reducer) {
-		return transform("Grouped Reduce", getType(), new StreamGroupedReduce<T>(
-				clean(reducer), keySelector));
+		return transform("Grouped Reduce", getType(), new StreamGroupedReduce<>(clean(reducer), keySelector));
```
Comment (ktzoumas) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: "Grouped Reduce" or simply "Reduce"?

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java [hunk 3]
```diff
@@ -87,7 +207,7 @@ public GroupedDataStream(DataStream<T> dataStream, KeySelector<T, KEY> keySelect
 		TypeInformation<R> outType = TypeExtractor.getFoldReturnTypes(clean(folder), getType(),
 				Utils.getCallLocationName(), true);
 
-		return transform("Grouped Fold", outType, new StreamGroupedFold<T, R>(clean(folder),
+		return transform("Grouped Fold", outType, new StreamGroupedFold<>(clean(folder),
 				keySelector, initialValue));
```
Comment (ktzoumas) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: "Grouped Fold" or simply "Fold"?

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/datastream/IterativeDataStream.java (hunk 1 above)
Comment (aljoscha) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/datastream/IterativeDataStream.java]: You're right, I'll also rename those.

**Comment by StephanEwen** [comment]
Impressive work, looks good!

Merging this means that we need to commit to reworking the `join()` implementation very soon.

If that is the case, +1 from my side

**Comment by ktzoumas** [comment]
+1 to merge. This will make the testing of the new API much easier

**Pull request closed without merging**